creator/data_reviews/signals.py

```diff
@@ -0,0 +1,98 @@
+from django.dispatch import receiver
+from django.db.models.signals import pre_delete, post_save
+
+from creator.files.models import Version
+from creator.events.models import Event
+from creator.data_reviews.models import DataReview, State
+
+
+@receiver(pre_delete, sender=Version)
+def file_version_pre_delete(sender, instance, using, *args, **kwargs):
+    """
+    When a file Version is deleted, update related DataReview states and fire
+    Data Review update events
+
+    Scenario 1
+    -----------------------------------
+    Create Study - fa.v1, fb.v1, fc.v1
+    Start DataReview - fa.v1, fb.v1
+    Version deleted - fc.v1
+
+    Is version part of any Data Review
+    No - do nothing
+
+    Scenario 2 - continuation of 1
+    -----------------------------------
+    Study - fa.v1, fb.v1
+    DataReview - fa.v1, fb.v1
+    Version deleted - fa.v1
+
+    Is version part of any Data Review
+    Yes - update all related review's state, emit event
+    """
+    for dr in instance.data_reviews.all():
+        if dr.state == State.WAITING:
+            dr.receive_updates()
+            dr.save()
+        Event(
+            study=dr.study,
+            file=instance.root_file,
+            user=dr.creator,
+            data_review=dr,
+            description=(
+                "File version {instance.pk} was deleted from data review "
+                f"{dr.pk}"
+            ),
+            event_type="DR_UPD",
+        ).save()
+
+
+@receiver(post_save, sender=Version)
+def file_version_post_save(sender, instance, created, *args, **kwargs):
+    """
+    When a file Version is created, update the states of the related
+    DataReview and fire events
+
+    Scenario 1
+    -----------------------------------
+    Create Study - fa.v1, fb.v1, fc.v1
+    Start DataReview - fa.v1, fb.v1
+    Receive new version - fc.v2
+
+    Check if instance.root_file has a version that is part of any Data Review
+    No - do nothing
+
+    Scenario 2 - continuation of 1
+    -----------------------------------
+    Study - fa.v1, fb.v1, fc.v2
+    DataReview - fa.v1, fb.v1
+    Receive new version - fa.v2
+
+    Check if instance.root_file has a version that is part of any Data Review
+    Yes - update all related review's state, emit event
+    """
+    # Only process new versions with root files
+    if (not created) or (not instance.root_file):
+        return
+
+    # For data reviews involving the root file of the version that was uploaded
+    # update the reviews' state + emit events
+    for dr in DataReview.objects.filter(
```

Comment:
Why can't you do the following here?
```
for dr in instance.data_reviews.filter(state=State.WAITING).all():
```

Reply:
My thought was even if the dr state doesn't change, we may still want to emit an event to alert the user that a new version was uploaded, if that new version is linked to a file in the review. I imagine a couple scenarios. The first is the obvious one where feedback has been sent to the investigator and we're waiting for updates. When we receive the updates we definitely want to make it known to the user that updates have been received. But then another scenario is, maybe the review is still IN_REVIEW and an analyst makes an update to a file (perhaps its something easy they can do without input from investigator, etc). I think we still want to emit this event to make it obvious that a change has been made to a file in the review.  

Perhaps what we could do is emit the event only if the review is not in closed or completed state? However, I would ask - do we actually still want to emit this event even if the review is closed/completed? I can imagine still getting updates to files even after the review is closed or completed and wanting some sort of record of this?

Reply:
Thinking some more..I have to go back and look at what that Data Review detail view will look like. If the event timeline will already have events about new versions being uploaded (like the File detail page) then maybe this event is redundant?

Reply:
Ok I think I understand the logic here better now. I would say this is good, except for you probably want to only emit events for data reviews that are not in terminal states. It would be strange to see a Data Review marked as approved or closed and then have events following that.